sequencer/src/auction.rs

```diff
@@ -0,0 +1,312 @@
+}
+
+/// Uniquely identifies an auction for sequencing rights of namespaces in the network
+#[derive(Clone, Copy)]
+struct AuctionId(u64);
+
+/// Uniquely identifies one auction phase for a specific auction
+#[derive(Clone, Copy)]
+struct AuctionPhaseId(AuctionId, AuctionPhaseKind);
+
+/// Describes one auction phase for a specific auction
+#[derive(Clone, Copy)]
+struct AuctionPhase {
+    id: AuctionPhaseId,
+    kind: AuctionPhaseKind,
+    start: ViewNumber,
+    end: ViewNumber,
+}
+
+/// Describes the 3 kinds of phases an active auction can be in
+#[derive(Clone, Copy, Eq, PartialEq)]
+enum AuctionPhaseKind {
+    Bid,
+    Assign,
+    Sequence,
+}
+
+#[derive(Debug, Default, Clone, Copy, Eq, PartialEq, Deserialize, Serialize, Hash)]
+pub struct Slot(u64);
+#[derive(Debug, Clone, Eq, PartialEq, Deserialize, Serialize, Hash)]
+pub enum FullNetworkTx {
+    Bid(BidTx),
+}
+
+impl FullNetworkTx {
+    pub fn execute(
+        &self,
+        state: &mut ValidatedState,
+    ) -> Result<(), (ExecutionError, FullNetworkTx)> {
+        match self {
+            Self::Bid(bid) => bid.execute(state),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Deserialize, Serialize, Hash)]
+pub struct BidTx {
+    body: BidTxBody,
+    signature: Signature,
+}
+
+/// A transaction to bid for the sequencing rights of a namespace
+#[derive(Debug, Clone, Eq, PartialEq, Deserialize, Serialize, Hash)]
+pub struct BidTxBody {
+    /// Account responsible for the signature
+    account: FeeAccount,
+    /// Fee to be sequenced in the network.  Different than the bid_amount fee
+    // FULL_NETWORK_GAS * MINIMUM_GAS_PRICE
+    gas_price: FeeAmount,
+    /// The bid amount designated in Wei.  This is different than
+    /// the sequencing fee (gas price) for this transaction
+    bid_amount: FeeAmount,
+    // TODO What is the correct type?
+    /// The public key of this sequencer
+    public_key: SequencerKey,
+    /// The URL the HotShot leader will use to request a bundle
+    /// from this sequencer if they win the auction
+    url: Url,
+    /// The slot this bid is for
+    slot: Slot,
+    /// The set of namespace ids the sequencer is bidding for
+    bundle: Vec<NamespaceId>,
+}
+
+// TODO consider a committable derive macro
+impl Committable for BidTxBody {
+    fn tag() -> String {
+        "BID_TX".to_string()
+    }
+
+    fn commit(&self) -> Commitment<Self> {
+        let comm = committable::RawCommitmentBuilder::new(&Self::tag())
+            .fixed_size_field("account", &self.account.to_fixed_bytes())
+            .fixed_size_field("gas_price", &self.gas_price.to_fixed_bytes())
+            .fixed_size_field("bid_amount", &self.bid_amount.to_fixed_bytes())
+            .var_size_field("url", self.url.as_str().as_ref())
+            .u64_field("slot", self.slot.0)
+            .var_size_field(
+                "bundle",
+                // TODO what is the correct way to serialize `Vec<NamespaceId>`
```

Comment:
@ggutoski do you know the answer?

Reply:
This code uses bincode via serde. I see no problem with this. Not sure why you need `as_slice()` here though.

Reply:
I don't! thanks for the feedback!